compiler/src/dotty/tools/dotc/transform/RenameLifted.scala

```diff
@@ -0,0 +1,46 @@
+package dotty.tools.dotc.transform
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Decorators._
+import dotty.tools.dotc.core.DenotTransformers.SymTransformer
+import dotty.tools.dotc.core.Flags._
+import dotty.tools.dotc.core.NameKinds._
+import dotty.tools.dotc.core.Names._
+import dotty.tools.dotc.core.Phases
+import dotty.tools.dotc.core.SymDenotations.SymDenotation
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.transform.TreeTransforms.MiniPhaseTransform
+
+/** Renames lifted classes to local numbering scheme */
+class RenameLifted extends MiniPhaseTransform with SymTransformer { thisTransformer =>
+
+  override def phaseName = "renameLifted"
+
+  override def runsAfterGroupsOf: Set[Class[_ <: Phases.Phase]] = Set(classOf[RestoreScopes])
+
+  def transformSym(ref: SymDenotation)(implicit ctx: Context): SymDenotation =
+    if (needsRefresh(ref.symbol)) ref.copySymDenotation(name = refreshedName(ref.symbol))
+    else ref
+
+  /** If the name of the symbol with a unique name needs to be refreshed
+   *    - if it is a lifted class
+   *    - if it is a lifted method
+   */
+  private def needsRefresh(sym: Symbol)(implicit ctx: Context): Boolean =
+    (sym.isClass || sym.is(Private | Method | JavaStatic)) && sym.name.is(UniqueName)
+
+  /** Refreshes the number of the name based on the full name of the symbol */
+  private def refreshedName(sym: Symbol)(implicit ctx: Context): Name = {
+    lazy val rewriteUnique: PartialFunction[Name, Name] = {
```

Comment:
Why lazy?

Reply:
Oh I see, that's a funny way to write recursion :smile:

Reply:
Do you know a cleaner way to define recursive partial functions?

Reply:
I would do it with a `def`

Reply:
Changed it. Though now it will create a new instance of the function each iteration, I guess this should not happen too often.